launch/launch/utilities/type_utils.py

```diff
@@ -0,0 +1,565 @@
+            if is_substitution(x) else float(cast(Union[int, float], x)) for x in value
+        ])
+    # Invalid input
+    raise TypeError(
+        "Input value is not an acceptable 'SomeValueType'."
+        f"Got value='{value}'"
+    )
+
+
+def is_normalized_substitution(x):
+    """
+    Return `True` if `x` is a normalized substitution.
+
+    A normalized substitution is a list with :py:class:`launch.Substitution` instances as items.
+    """
+    return (
+        isinstance(x, list) and
+        len(x) > 0 and
+        all(isinstance(item, Substitution) for item in x)
+    )
+
+
+def perform_typed_substitution(
+    context: LaunchContext,
+    value: NormalizedValueType,
+    data_type: Optional[AllowedTypesType]
+) -> AllowedValueType:
+    """
+    Perform a normalized typed substitution.
+
+    This function becomes handy when you need a substitution which result will be coerced to a
+    specific type.
+
+    See :py:obj:`normalize_typed_substitution` for an example.
+
+    :param context: A :py:class:`launch.LaunchContext` instance.
+        It will be used to perform the substitutions in `value`, if there are any.
+    :param value: normalized typed substitution to be performed.
+    :param data_type: The return value will satisfy `is_instance_of(output) == True`.
+        If after trying to coerce the performed substitution that isn't achieved,
+        `ValueError` is raised.
+        See :py:func:`is_instance_of`.
+    :return: the result of performing all the substitutions in value and coercing the result.
+    :raise: `TypeError` if the normalized `value` cannot later be resolved to an instance
+        of `data_type`, or to a valid type when `data_type is `None`.
+    :raise: `ValueError` if `data_type` is not valid.
+        See :py:obj:`AllowedTypesTuple`.
+    """
+    if isinstance(value, ScalarTypesTuple):
+        if data_type is not None and not is_instance_of(value, data_type):
+            raise TypeError(
+                f'value=`{value}` is a scalar and not an instance of `{data_type}`')
+        return value
+    elif is_normalized_substitution(value):
+        return coerce_to_type(
+            perform_substitutions(context, cast(List[Substitution], value)),
+            data_type
+        )
+    elif isinstance(value, list):
+        scalar_type: Optional[ScalarTypesType] = None
+        if data_type is not None:
+            scalar_type, is_list = extract_type(data_type)
+            if not is_list:
+                raise ValueError(
+                    'The input value is a list, cannot convert it to a scalar. '
+                    f"Got value='{value}', expected type {scalar_type}."
+                )
+        output = [
+            coerce_to_type(
+                perform_substitutions(context, cast(List[Substitution], x)), scalar_type)
+            if is_normalized_substitution(x) else x for x in value
+        ]
+        is_instance_of(output, data_type)
```

Comment:
@ivanpauno perhaps you wanted:
```suggestion
        assert is_instance_of(output, data_type)
```
instead? There's no side-effect to `is_instance_of` (nor should be IMO).

Reply:
https://github.com/ros2/launch/pull/438/commits/cd3fafeca6102d69dc480c3a887ea74a2e659e8a